### ISSUE TITLE
[Prep] Set up SDK & reusable map view component

### ISSUE THREAD

**Issue by Johennes** [vector-im/element-meta#1684]
# Size estimate

M

# Dependencies

* None

# Context

We will rely on some external services to implement certains aspects of the location sharing feature.
Element has a paid subscription with maptiler.com which offers APIs to enable all the external services we need:
  * It has a tile server for maplibre-native to connect to via style.json URLs.
  * It has a static map service.
  * It has a geocoding service.

Our subscription gives us a single API key we can use with all these services (by appending it to their respective URLs).

### Interactive map
* It's a map the user can zoom in/out drag and pan.
* Its view is managed by the maplibre-native sdk which in turn connects to a tile server to get the necessary data.

### Static map
* It's a non-interactive a map.
* It's downloaded as an image from a web service that returns a .png (or other formats) given some input params (e.g. lat/lon/zoom/pixel dimension...).

### Reverse geocoding
* It's the process of turning a lat/lon pair into a street address.
* It is handled via a web service that returns an address string given some input params (e.g. lat/lon).

## Relevant MSCs
- https://github.com/matrix-org/matrix-spec-proposals/blob/matthew/location/proposals/3488-location.md

## Relevant dicussions:
- https://matrix.to/#/!kCCQTCfnABLKGGvQjo:matrix.org/$6mJfRYdCMD7Cw0IXHFZBwvlbx274W7hLkdrYO4oIyEQ?via=matrix.org&via=element.io&via=one.ems.host

# Acceptance criteria

## Interactive map
* A SwiftUI/Compose wrapper is created that:
  * Wraps the [maplibre-native](https://github.com/maplibre/maplibre-native) SDK's MapView to connect to Maptiler's tile server.
  * Uses the light or dark map `style.json` URLs depending on the UI's light/dark state.
  * Uses our Maptiler subscription's api key.
  * Draws a pin over the desired location.

## Static map
* A SwiftUI/Compose wrapper is created that:
  * Loads the map image via network using [Maptiler's static image http api](https://docs.maptiler.com/cloud/api/static-maps/).
  * Uses the light or dark static map URL depending on the UI's light/dark state.
  * Uses our Maptiler subscription's api key.
  * Draws a pin over the desired location (which should be at the center of the image).

## Reverse geocoding
* Business logic is created that:
  * Calls into [Maptiler's geocoding API](https://docs.maptiler.com/cloud/api/geocoding/#search-by-coordinates-reverse) to get an address from a lat/lon pair.
    * Returns the content of the json field `features[0].place_name` of the response body.
  * Uses our Maptiler subscription's api key.

## General config
* API keys and map URLs are available at: https://docs.google.com/document/d/1T048whP7KNaJlbUma1uSe6LZ6GECQSypC_VnUbmzw88/edit?usp=sharing

# Out of scope
* Location permission handling – https://github.com/vector-im/element-meta/issues/1674
* Dark mode – https://github.com/vector-im/element-meta/issues/1683

# Future work
* Customise maps provider - https://github.com/vector-im/element-meta/issues/1722

# Subtasks

```[tasklist]
### Android
- [ ] https://github.com/vector-im/element-x-android/issues/688
```

```[tasklist]
### iOS
- [ ] https://github.com/vector-im/element-x-ios/issues/1003
- [ ] https://github.com/vector-im/element-x-ios/issues/1014
- [ ] https://github.com/vector-im/element-x-ios/issues/1015
```

```[tasklist]
### Other
```

**Comment by julioromano** [comment]
**TLDR: The following info has been integrated in the issue description.**

Some info gathered from EA app:

```
MAP_BASE_URL = "https://api.maptiler.com/maps/streets/style.json"
STATIC_MAP_BASE_URL = "https://api.maptiler.com/maps/basic/static/"
DEFAULT_PIN_ID = "DEFAULT_PIN_ID"
INITIAL_MAP_ZOOM_IN_PREVIEW = 15.0
INITIAL_MAP_ZOOM_IN_TIMELINE = 17.0
```

- The main map url comes from `.well-known` and api key is not added to it programmatically so I believe the server returns an URL with built in api key (we should double check this).
- There's an hardcoded fallback tile server url which we use with an api key. This is used whenever there is no conf from `.well-known`.
- Static (non-vector) maps are always created using the hardcoded static map base url (so this doesn't go through `.well-known`. We use the same api key as above with this url.
- After a very quick look, it doesn't seem the rust sdk exposes access to `.well-known` yet.

Out of scope:
- Dark mode: 
  - Using the Maptiler dashboard one can easily create new map stiles (such as a dark one), Callum has already done it.
  - The new map style implies having a second map url for dark more. So we can easily add light/dark mode urls for the hardcoded urls, if there's an override url in `.well-known` we can just use that in place of both the built in light and dark variants (until we offer a dark url from within `.well-known`.
  - See: https://www.maptiler.com/news/2020/05/maps-for-dark-mode/
